torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -281,27 +281,41 @@ def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
         # Synchronize the detected inf across the ranks
         optimizer_state = self._per_optimizer_states[id(optimizer)]
         future_handles = []
-
-        for v in optimizer_state["found_inf_per_device"].values():
-            if v.device.type == "cpu":
-                v_on_cuda = v.cuda()
+        found_inf_on_cudas = []
+
+        for found_inf in optimizer_state["found_inf_per_device"].values():
+            if found_inf.device.type == "cpu":
+                # found_inf is a scalar tensor on CPU
+                # it's small enough to live in pinned memory
+                # to speed up H2D transfer
+                found_inf_on_cuda = found_inf.pin_memory().cuda()
+                found_inf_on_cudas.append(found_inf_on_cuda)
                 future_handles.append(
                     dist.all_reduce(
-                        v_on_cuda, async_op=True, group=self.process_group
+                        found_inf_on_cuda, async_op=True, group=self.process_group
```

Comment:
Is the reason that this all-reduce is async so that we can overlap with H2D copies? I am not exactly sure since the H2D copies look like they are blocking, so the CPU will not issue the next all-gather until the previous H2D copy finished (IIUC).

Reply:
discussed offline async all-reduce brings value that cpu thread can keep doing H2D -> H2D copies without get blocked on all-reduce (H2D -> all-reduce -> H2D -> all-reduce ...)

Reply:
In particular, the thinking is that we can do: H2D -> all-reduce (async) -> H2D, where the all-reduce overlaps with the _subsequent_ H2D but not any others. In that sense, the waiting of all all-reduce handles at the end is only for convenience; in reality, the next `.cuda()` already blocks the CPU thread until the H2D copy finishes before it can issue the next all-reduce.

Given that these copies and all-reduces operate on only a single element at a time, I am not sure how much gain we get from overlapping here.